intermediate_source/realtime_rpi.rst

```diff
@@ -0,0 +1,259 @@
+
+  $ python3 -c "import torch; print(torch.__version__)"
+  1.10.0+cpu
+
+.. image:: https://user-images.githubusercontent.com/909104/152874271-d7057c2d-80fd-4761-aed4-df6c8b7aa99f.png
+
+
+Video Capture
+~~~~~~~~~~~~~~
+
+For video capture we're going to be using OpenCV to stream the video frames
+instead of the more common ``picamera``. `picamera` isn't available on 64-bit
+Raspberry Pi OS and it's much slower than OpenCV. OpenCV directly accesses the
+``/dev/video0`` device to grab frames.
+
+The model we're using (MobileNetV2) takes in image sizes of ``224x224`` so we
+can request that directly from OpenCV at 36fps. We're targeting 30fps for the
+model but we request a slightly higher framerate than that so there's always
+enough frames.
+
+.. code:: python
+
+  import cv2
+  from PIL import Image
+
+  cap = cv2.VideoCapture(0)
+  cap.set(cv2.CAP_PROP_FRAME_WIDTH, 224)
+  cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 224)
+  cap.set(cv2.CAP_PROP_FPS, 36)
+
+OpenCV returns a ``numpy`` array in BGR so we need to read and do a bit of
+shuffling to get it into the expected RGB format.
+
+.. code:: python
+
+    ret, image = cap.read()
+    # convert opencv output from BGR to RGB
+    image = image[:, :, [2, 1, 0]]
+
+NOTE: You can get even more performance by training the model directly with OpenCV's BGR data format to remove the conversion step.
+
+Image Preprocessing
+~~~~~~~~~~~~~~~~~~~~
+
+We need to take the frames and transform them into the format the model expects. This is the same processing as you would do on any machine with the standard torchvision transforms.
+
+.. code:: python
+
+    from torchvision import transforms
+
+    preprocess = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+    ])
+    input_tensor = preprocess(image)
+    input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model
+
+MobileNetV2: Quantization and JIT
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For optimal performance we want a model that's quantized and fused. Quantized
+means that it does the computation using int8 which is much more performant than
+the standard float32 math. Fused means that consecutive operations have been
+fused together into a more performant version where possible. Commonly things
+like activations (``ReLU``) can be merged into the layer before (``Conv2d``)
+during inference.
+
+The aarch64 version of pytorch requires using the ``qnnpack`` engine.
+
+.. code:: python
+
+    import torch
+    torch.backends.quantized.engine = 'qnnpack'
+
+For this example we'll use a prequantized and fused version of MobileNetV2 that's provided out of the box by torchvision.
+
+.. code:: python
+
+    from torchvision import models
+    net = models.quantization.mobilenet_v2(pretrained=True, quantize=True)
+
+We then want to jit the model to reduce Python overhead and fuse any ops. Jit gives us ~30fps instead of ~20fps without it.
```

Comment:
Kinda interesting that vanilla pytorch worked fine. curious if you tried using pytorch live

Reply:
I think PyTorch Live is only for iOS/Android https://pytorch.org/live/docs/tutorials/get-started/ -- RPi is just aarch64 Linux so standard PyTorch works out of the box